File: Week7_DFS/isSameTree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        /* If both nodes are null, then tree structure is identical */
        if (p == null && q == null) {
            return true;
        }

        /* If 1 of 2 nodes is null (other has value), tree structure differs */
        if (p == null || q == null) {
            return false;
        }

        /* If value of nodes are different, then they aren't the same */
        if (p.val != q.val) {
            return false;
        }

        /* Recurse to left and right and check. && Used since all nodes must be equal. */
        return (isSameTree(p.left, q.left) && isSameTree(p.right, q.right));
    }
}

File: Session5_DFS/isSameTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        # If both nodes are null, then tree structure is identical
        if not p and not q:
            return True
        # If 1 of 2 nodes is null (other has value), tree structure differs
        if not p or not q:
            return False
        
        # If value of nodes are different, then they aren't the same
        if p.val != q.val:
            return False
        
        # Recurse to left and right and check. && Used since all nodes must be equal.
        return (self.isSameTree(p.left, q.left) and
            self.isSameTree(p.right, q.right))